### PR TITLE
[MM-41905] Adding more hooks and optimizing functional components

### PR DESCRIPTION
**Pull request by BenCookie95**
#### Summary
Fixing most of the comments from here https://github.com/mattermost/mattermost-webapp/pull/9434

- Made more use of useCallback and React.memo
- Made another small component for the view group modal header

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-41905

#### Release Note

```release-note
NONE
```

### CODE REVIEW

File: components/add_users_to_group_modal/add_users_to_group_modal.tsx [hunk 1]
```diff
@@ -71,12 +71,24 @@ const AddUsersToGroupModal = (props: Props) => {
         } else {
             goBack();
         }
-    };
+    }, []);
```
Comment (jespino) [inline components/add_users_to_group_modal/add_users_to_group_modal.tsx:74]: I miss here the `goBack` dependency

Comment (jespino) [reply components/add_users_to_group_modal/add_users_to_group_modal.tsx:74]: and maybe, but I don't think is going to change, the `props.actions.addUsersToGroup`

File: components/update_user_group_modal/update_user_group_modal.tsx [hunk 2]
```diff
@@ -62,16 +62,16 @@ const UpdateUserGroupModal = (props: Props) => {
         setName(value);
         setHasUpdated(true);
         setMention(newMention);
-    };
+    }, []);
```
Comment (jespino) [inline components/update_user_group_modal/update_user_group_modal.tsx:65]: I miss the `mention` dependency here.

File: components/update_user_group_modal/update_user_group_modal.tsx [hunk 3]
```diff
@@ -115,11 +115,12 @@ const UpdateUserGroupModal = (props: Props) => {
         } else {
             goBack();
         }
-    };
-    const goBack = () => {
+    }, []);
```
Comment (jespino) [inline components/update_user_group_modal/update_user_group_modal.tsx:118]: I miss here  `name`, `mention`, `goBack`, `props.groupId` and maybe `props.actions.patchGroup`

File: components/view_user_group_modal/view_user_group_header_sub_menu/view_user_group_header_sub_menu.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+        actions.openModal({
+            modalId: ModalIdentifiers.EDIT_GROUP_MODAL,
+            dialogType: UpdateUserGroupModal,
+            dialogProps: {
+                groupId: group.id,
+                backButtonCallback: backButtonAction,
+            },
+        });
+        onExited();
+    }, [actions.openModal, group.id, backButtonAction]);
```
Comment (jespino) [inline components/view_user_group_modal/view_user_group_header_sub_menu/view_user_group_header_sub_menu.tsx:60]: missing here the `onExited` dependency.

File: components/view_user_group_modal/view_user_group_header_sub_menu/view_user_group_header_sub_menu.tsx [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+    const leaveGroup = useCallback(async () => {
+        await actions.removeUsersFromGroup(group.id, [currentUserId]).then(() => {
+            decrementMemberCount();
+        });
+    }, [group.id, actions.removeUsersFromGroup, decrementMemberCount]);
```
Comment (jespino) [inline components/view_user_group_modal/view_user_group_header_sub_menu/view_user_group_header_sub_menu.tsx:66]: missing here the `currentUserId` dependency.

File: components/view_user_group_modal/view_user_group_header_sub_menu/view_user_group_header_sub_menu.tsx [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    const joinGroup = useCallback(async () => {
+        await actions.addUsersToGroup(group.id, [currentUserId]).then(() => {
+            incrementMemberCount();
+        });
+    }, [group.id, actions.addUsersToGroup, incrementMemberCount]);
```
Comment (jespino) [inline components/view_user_group_modal/view_user_group_header_sub_menu/view_user_group_header_sub_menu.tsx:72]: Missing the currentUserId dependency here.

**Review by jespino — CHANGES_REQUESTED** [review]
Detected some dependencies missing in the hooks, but other than that looks good to me.

**Review by jespino — APPROVED** [review]
LGTM. Thanks for the effort :)

**Review by mkraft — APPROVED** [review]

**Comment by BenCookie95** [comment]
@furqanmlk For testing we just need to make sure nothing is broken, there should be no functional changes

**Review by furqanmlk — APPROVED** [review]
Thanks @BenCookie95 
sorry for the delay e2e build was broken. 
E2E is passed ✅ 

https://automation-dashboard.vercel.app/cycles/353321ea-4a6c-4c00-b82d-1ffe426ab081

**Pull request merged**